### ISSUE TITLE
Unable to evaluate in Blazor Wasm

### ISSUE THREAD

**Issue by mkArtakMSFT** [mono/mono#19942]
---
**Issue moved from dotnet/aspnetcore#22569**
- Please respond to @srudin.

---

_From @srudin on Friday, June 5, 2020 9:18:30 AM_

### Describe the bug
Debugging generally works, my breakpoints are hit. However sometimes I cannot display all or some of the variables: The locals window, quick watch etc. simply display "Unable to evaluate" for all variables, objects etc. Therefore expanding an object (by clicking on the +) also doesn't work. Debugging without seeing any values is of course not very helpful...

I am using the newest releases of both VS2019 (16.6.1) and the .NET SDK (3.1.300). Is there something else I need to do?

### To Reproduce
Create a new client side Blazor app, set a breakpoint, debug.

@guardrex

**Comment by mkArtakMSFT** [comment]
---
**Issue moved from dotnet/aspnetcore#22569**
- Please respond to @mkArtakMSFT.

---

_From @mkArtakMSFT on Tuesday, June 9, 2020 4:56:55 PM_

Thanks for contacting us.
If you would have to categorize, which are the most common things / expressions you'd expect to be evaluated which currently don't?

**Comment by radical** [comment]
@srudin Could you please record a screencast to show the issue?

**Comment by dme-development** [comment]
I'm having the exact same issue as @dave-schneider-ila. When inheriting the razor view from a modal derived from ComponentBase it is impossible to see the values of any variable while debugging (Unable to evaluate). 

@radical : I believe this explanation is good enough to investigate without screencast.

**Comment by radical** [comment]
@srudin @dme-development I haven't been able to reproduce this. Could you share a simple project which hits this issue? And please mention the versions of VS, and Blazor that you are using now.

**Comment by Postlagerkarte** [comment]
@radical :  I get "Children could not be evaluated"

![image](https://user-images.githubusercontent.com/8482283/89825187-7e3ec480-db54-11ea-9da7-8d7c82a1a917.png)

Is this related or should I open a separate issue?

**Comment by ehsangfl** [comment]
I Have this problem too

**Comment by radical** [comment]
@Postlagerkarte yes, please could you open a separate issue for this? thanks!

**Comment by radical** [comment]
@srudin @dme-development Are you still seeing this issue, with preview8? You won't be able to see inherited members on object yet.

**Comment by radical** [comment]
@Postlagerkarte https://github.com/dotnet/runtime/pull/41559

**Comment by JeffreyYeTHU** [comment]
Hey@radical I am running into the same error when calling a Web API using httpClient.GetAsync("serverUrl") in a Blazor web assembly app. I am using VS 16.7.2,  TargetFramework netstandard2.1, RazorLangVersion 3.0. I have performed the following steps to pinpoint the issue:
1. The backend API behaves normally when calling from PostMan.
2. I create a brand new aspnet core web app, just to call the API, works fine.
So I suspect the issue comes from the Blazor web assembly app part. Does this deserve a new Issue? I will try the preview version of the visual studio and dotnet 5 to see if the same problem will occur.

**Comment by JeffreyYeTHU** [comment]
Tried dotnet 5 with visual studio 16.8.0 preview 2.1:
1. Blazor server app calls web API ok
2. Blazor wasm app calls web API with error:
![image](https://user-images.githubusercontent.com/28821209/91971915-87cedd00-ed4c-11ea-8bfe-602592e59148.png)

**Comment by radical** [comment]
@JeffreyYeTHU this issue is about the debugger not being able to show some values. You should file a separate issue for the webapi not working.

**Comment by JeffreyYeTHU** [comment]
@radical Thanks for reply. I try another PC with windows using Blazor wasm (Net core 3.1, VS 16.7.2), API call works. So I am not sure about the exact issue. I will experiment more and if the issue is reproducible, I will file the issue.

**Comment by AoiMasa** [comment]
@radical  
I have the same issue: it hits breakpoints but cannot evaluate variables.  It seems to occurs only when the code is in a separate class.

**How to reproduce:** 
Create a new client side Blazor app
Create CounterBase class with Counter code
Debug CounterBase IncrementCount method
Result:
![BlazorDebugBug](https://user-images.githubusercontent.com/18733930/92742906-5a9fb180-f380-11ea-960c-97fdbeccae41.PNG)

----------
.NET core 3.1
VS 2019 16.7.3

**Comment by radical** [comment]
@AoiMasa could you please share that project?

**Comment by AoiMasa** [comment]
@radical 
Here you go
[BlazorAppDummy.zip](https://github.com/mono/mono/files/5207238/BlazorAppDummy.zip)

**Comment by bfmsoft** [comment]
I am seeing the same thing.  Can hit breakpoints but no variables are evaluated.  I am on .Net 5 RC1

![image](https://user-images.githubusercontent.com/59542880/95026702-38d4da00-0661-11eb-8bd6-a8b71e54c0d9.png)

**Comment by johannesmols** [comment]
I'm also having the same issue, I'm already on .NET 5.0.0
![image](https://user-images.githubusercontent.com/11809436/98973610-3f822700-2514-11eb-90e5-861df55fb21c.png)

**Comment by coeur85** [comment]
same here

**Comment by angelalaweai** [comment]
having same issues with blazor wasm. unable to evaluate in watch window, I have having to convert everything to json and put it out with a console.writeln to the dev tools for the browser to view information

![image](https://user-images.githubusercontent.com/62156838/101121303-cf6c3b80-35bd-11eb-831d-ff53cec54535.png)


version I am running in vs 2019
Microsoft Visual Studio Professional 2019
Version 16.7.7
VisualStudio.16.Release/16.7.7+30621.155
Microsoft .NET Framework
Version 4.8.04084

Installed Version: Professional

ASP.NET and Web Tools 2019   16.7.532.28833
ASP.NET and Web Tools 2019

ASP.NET Core Razor Language Services   16.1.0.2035807+72d099b977d3a85e65fa3b0614ca8cfc803fef02
Provides languages services for ASP.NET Core Razor.

ASP.NET Web Frameworks and Tools 2019   16.7.532.28833
For additional information, visit https://www.asp.net/

Azure App Service Tools v3.0.0   16.7.532.28833
Azure App Service Tools v3.0.0

Azure Functions and Web Jobs Tools   16.7.532.28833
Azure Functions and Web Jobs Tools

C# Tools   3.7.0-6.20514.1+600c0dadd199296df3ed07dfabdc014563a77f7c
C# components used in the IDE. Depending on your project type and settings, a different version of the compiler may be used.

Common Azure Tools   1.10
Provides common services for use by Azure Mobile Services and Microsoft Azure Tools.

Cookiecutter   16.7.20248.6
Provides tools for finding, instantiating and customizing templates in cookiecutter format.

IntelliCode Extension   1.0
IntelliCode Visual Studio Extension Detailed Info

Microsoft Azure Tools   2.9
Microsoft Azure Tools for Microsoft Visual Studio 2019 - v2.9.30701.1

Microsoft Continuous Delivery Tools for Visual Studio   0.4
Simplifying the configuration of Azure DevOps pipelines from within the Visual Studio IDE.

Microsoft JVM Debugger   1.0
Provides support for connecting the Visual Studio debugger to JDWP compatible Java Virtual Machines

Microsoft Library Manager   2.1.79+ge3567815aa.RR
Install client-side libraries easily to any web project

Microsoft MI-Based Debugger   1.0
Provides support for connecting Visual Studio to MI compatible debuggers

Microsoft Visual Studio Tools for Containers   1.1
Develop, run, validate your ASP.NET Core applications in the target environment. F5 your application directly into a container with debugging, or CTRL + F5 to edit & refresh your app without having to rebuild the container.

MySQL for Visual Studio   1.2.9
Data design and management tools for MySQL. Copyright (c) 2007, 2019, Oracle and/or its affiliates. All rights reserved.

Node.js Tools   1.5.20701.1 Commit Hash:b0da3724234b78593207434292d9f891440edea1
Adds support for developing and debugging Node.js apps in Visual Studio

NuGet Package Manager   5.7.0
NuGet Package Manager in Visual Studio. For more information about NuGet, visit https://docs.nuget.org/

ProjectServicesPackage Extension   1.0
ProjectServicesPackage Visual Studio Extension Detailed Info

Python   16.7.20248.6
Provides IntelliSense, projects, templates, debugging, interactive windows, and other support for Python developers.

Python - Conda support   16.7.20248.6
Conda support for Python projects.

Python - Django support   16.7.20248.6
Provides templates and integration for the Django web framework.

Python - IronPython support   16.7.20248.6
Provides templates and integration for IronPython-based projects.

Python - Profiling support   16.7.20248.6
Profiling support for Python projects.

SQL Server Data Tools   16.0.62008.03130
Microsoft SQL Server Data Tools

TypeScript Tools   16.0.20702.2001
TypeScript Tools for Microsoft Visual Studio

Visual Basic Tools   3.7.0-6.20514.1+600c0dadd199296df3ed07dfabdc014563a77f7c
Visual Basic components used in the IDE. Depending on your project type and settings, a different version of the compiler may be used.

Visual F# Tools 10.10.0.0 for F# 4.7   16.7.0-beta.20361.3+3ef6f0b514198c0bfa6c2c09fefe41a740b024d5
Microsoft Visual F# Tools 10.10.0.0 for F# 4.7

Visual Studio Code Debug Adapter Host Package   1.0
Interop layer for hosting Visual Studio Code debug adapters in Visual Studio

Visual Studio Container Tools Extensions (Preview)   1.0
View, manage, and diagnose containers within Visual Studio.

Visual Studio Tools for Containers   1.0
Visual Studio Tools for Containers

**Comment by radical** [comment]
More information is required to reproduce, and fix the issue. It would be useful to:

1. ensure you are on the latest .net 5.x release
2. open a separate issue with details of the code where you are seeing the issue. A small test project, if possible, to reproduce the issue would be extremely helpful.

This error message could represent different problems, so separate issues with details of the code would help.

**Comment by ehsangfl** [comment]
I try to provide more description,
because this problem occures when we want to see value of variable from type of a class on other assembly of solution, based on a class from thrid party!!!

Class A => On assembly Model.dll
Class B => On third party assembly, referenced to Model.dll
Class C => a property of class A. type C defined on third party
Class D => has property of type A,


```
public class A : B
{
    public C property1 {get;set;}

   public A()
 {
     property1  = new property1();
 }

}

```
 Usage of Class A on class D on assembly Views.dll
```

public class D
{
    public A property2 {get;set;}

private void TestFunctionForBreakPoint()
{
   this.Property2 = new A();
  var A =  A.Property1;
}

}


```
set breakpoint on var A = A.property1 and value of this property not shown.

I try to create sample project for you
BR

**Comment by radical** [comment]
@ehsangfl I have opened a new issue - https://github.com/dotnet/runtime/issues/45595

**Comment by radical** [comment]
@johannesmols @coeur85 Could you please file separate issues with more details about the code, and exactly what you are trying? "Unable to evaluate" is a generic message which just means that *something* failed. The reason for that, or the kinda code/variable/watch expression that might result in that can be completely different.

**Comment by ig16022** [comment]
Same issue here.

**Comment by aardill** [comment]
Is anything happening on this issue? I can't currently debug properly (need to use dotnet watch and console.writeline

**Comment by federicodibe** [comment]
Same problem here when I try to call .GetType()

The problem occurs both from the chrome debug console and from visual studio 2019 watch expression (version 16.10.0)

![unable](https://user-images.githubusercontent.com/62238562/124588854-92b93900-de59-11eb-8e7d-da7b0ceb872e.jpg)

**Comment by Luk164** [comment]
Same problem just occurred to me, did anyone find a workaround yet?

**Comment by bmcdonaldv** [comment]
I see this come and go.  Sometimes it will see variables at a breakpoint and sometimes not.  Other times breakpoints stop working.

**Comment by Luk164** [comment]
> 
> 
> I see this come and go. Sometimes it will see variables at a breakpoint and sometimes not. Other times breakpoints stop working.

Exactly, after a few tries it started working again. Still a bit spotty but I guess we should count our blessings. Hopefully this will be fixed soon.

**Comment by bmcdonaldv** [comment]
Some times it requires to restart VS.  I am using VS 2022 btw.

**Comment by Andrew-Webb** [comment]
I am always running into `Unable to evaluate (Unable to evaluate)` when debugging in VS 2019 on the client side of a Blazor WebAssembly project.  It happens every day.

@radical - I built a simple repro just now.

First of all: I am using Visual Studio Professional 2019, version 16.10.4

1) Create a new Blazor WebAssembly project, ASP.NET hosted, .NET 5.0.
2) Run it.
3) Go to the `Fetch data` page.  This will probably fail, because the template doesn't set the `.Server` project as the startup project.
4) Quit.  Set the `.Server` project as the startup project.  Restart.
5) Fetch data... there's your weather forecast data.  Nice.
6) Edit `FetchData.razor` and change the `GetFromJsonAsync` call to this...

```
        try
        {
            forecasts = await Http.GetFromJsonAsync<WeatherForecast[]>("WeatherForecast1");
        }
        catch (Exception ex)
        {
            Console.WriteLine(ex.Message);
        }
```

7) Note that the path is now `WeatherForecast1` - which should cause an error.
8) Re-run, fetch data.. there should be an error message in the console window.
9) Quit.  Put breakpoint on the `Console.WriteLine` line.  Restart.  (Sometimes an app errors on restart if breakpoints are set; in this case start without breakpoints, and add your breakpoint once the app is up and running.)
10) Go to fetch data page.  Debugger should have broken on the writeline line.
11) Do a QuickWatch on `ex`.  This should work.
12) We suspect that the type of the exception is not actually `System.Exception`, so in QuickWatch window type `ex.GetType()` and press enter.

Result --> 
![unabletoeval](https://user-images.githubusercontent.com/3470891/127134204-2bb7713a-1091-45b3-9321-eb6a3365bae0.png)

**Comment by aardill** [comment]
Looks like this issue should be fixed in .NET 6.0 as stated in the other thread referred to by @radical . Maybe this thread should also be closed then.
https://github.com/dotnet/runtime/issues/45595

**Comment by Andrew-Webb** [comment]
@aardill - the other thread is probably related, but the initial premise (where classes A, B, C, D are, and how they interrelate) seems very specific and complicated.  Our issue here is far simpler, just see my uber-simple repro two posts above.

Certainly our issue here in this thread _should_ be fixed in .NET 6.0... but it would be nice if it could have a milestone and assignees like the other issue, so we can be sure.  @SamMonoRT - can you give this issue some love?

To go into .NET 6.0 still being "Unable to evaluate" the simplest things would be very annoying.

**Comment by Andrew-Webb** [comment]
I have taken my simple repro above, and re-created it (with latest Visual Studio 2019 version) in the `dotnet/runtime` repo as https://github.com/dotnet/runtime/issues/59694

**Comment by Andrew-Webb** [comment]
@aardill  Ok, based on the issue I created in `dotnet/runtime` I am now convinced that this issue **will** be fixed in .NET 6.0.

I would say that this issue can now be closed, on the understanding that the fix is to migrate to .NET 6.0 (when fully released); there's no planned fix for .NET 5.0.

**Comment by radical** [comment]
cc @thaystg

**Comment by martinsmith-eapexecutive** [comment]
I am getting this problem in my .NET 6.0 project.
It is an existing project created in preview 3.0 since upgraded to 4.1, 
For a newly created project it is okay

**Comment by martinsmith-eapexecutive** [comment]
I figured out the solution, quite simple really 
Just update the references in the project file to this, as of now
		<PackageReference Include="Microsoft.AspNetCore.Components.WebAssembly" Version="6.0.0-rc.1.21452.15" />
		<PackageReference Include="Microsoft.AspNetCore.Components.WebAssembly.DevServer" Version="6.0.0-rc.1.21452.15" PrivateAssets="all" />

**Comment by Andrew-Webb** [comment]
> I'm running into the same problem with collections (Lists & Dictionaries)

@mvnoppen It might worth raising a highly specific issue wrt evaluating lists and dictionaries (and arrays?) on .NET 6.0 (at the indicated RC version) on https://github.com/dotnet/runtime  E.g. my issue https://github.com/dotnet/runtime/issues/59694 got a near instant response. @thaystg seems to be very dialled in to this problem, and hopefully will be able to confirm/deny the problem pretty quickly, and report back to the team if confirmed.

It's worth pushing to get these edge cases re. failure-to-evaluate cleared up before .NET 6.0 is released, I think.

**Comment by rajeshaz09** [comment]
This issue is not fixed in .NET6. Getting same problem with VS2022 and .NET6 Blazor WASM.

**Comment by thaystg** [comment]
Yes, this is fixed on main. For .net7 for sure. We will decide if we will backport to .net6 soon.

**Comment by rajeshaz09** [comment]
@thaystg  .NET6 is LTS. So please consider to backport.

**Comment by rajeshaz09** [comment]
@thaystg  May be I have different issue. .NET 5.x with VS2019 worked fine. This issue I observed in .NET 6 with VS2022. 
Any new changes in .NET6 and VS2022?

**Comment by thaystg** [comment]
What you are trying to evaluate? It's an array item as described in this issue?

**Comment by rajeshaz09** [comment]
@thaystg No.  Except injected services and parameters, I am unable to expand details of any other property.

**Comment by thaystg** [comment]
Can you please explain me how to reproduce it? Can you share your project?
Can you produce it in a Template of a Blazor WebAssembly app?

**Comment by rajeshaz09** [comment]
@thaystg  I figured out the reason. One of the dependent project which has model classes built using DebugType as Full instead of portable. Changed that and all the issues are resolved now

**Comment by bfmsoft** [comment]
I had this same issue.  Could you please consider have WASM projects check the DebugType and if not to correct one throw an error?  This has bitten me twice and now see another poor soul got this.

**Comment by rajeshaz09** [comment]
> I had this same issue. Could you please consider have WASM projects check the DebugType and if not to correct one throw an error? This has bitten me twice and now see another poor soul got this.

+1 
I wasted close to 2 days.

**Comment by thaystg** [comment]
I could reproduce it, I hope I can fix it soon!

**Comment by thaystg** [comment]
Thanks for the help to find what was going on.

**Comment by ChristopherHaws** [comment]
I also have this issue when trying to evaluate a dictionary:
![image](https://user-images.githubusercontent.com/5934505/152412126-20e8342c-53a6-4956-a2ea-d9f4ee419518.png)

Even if I try to materialize it as an array is still does not work:
![image](https://user-images.githubusercontent.com/5934505/152412236-a5d71948-ba49-4159-a2f6-f28a466334a9.png)

I am running on the latest versions of all my nuget packages and using SDK 6.0.101 in VS 2022.

**Comment by thaystg** [comment]
@ilonatommy , is it something that you have already fixed? If yes, which is the PR?
Thanks!!!

**Comment by ilonatommy** [comment]
> @ilonatommy , is it something that you have already fixed? If yes, which is the PR? Thanks!!!

Yes, I confirm that this feature is already fixed: 
![image](https://user-images.githubusercontent.com/32700855/152515453-54853447-4f0a-45f3-a00d-a204180b042f.png)

Sorry, I cannot find the PR but I checked on current version of main with HEAD on [here](https://github.com/dotnet/runtime/tree/bdae17fc0ec222e75e651d96af74e9667c01602e).

**Comment by EasyMod** [comment]
[ilonatommy](https://github.com/ilonatommy)
I have allready same problem, it happen when debagging blazor wasm+api and running api server. Even i run only blazor wasm client for debug then its ok

**Comment by mohaaron** [comment]
I am also seeing this happen in a larger project I'm working on. I'm not sure of the reason and am still trying to reproduce it in a small project.

**Comment by thaystg** [comment]
@lewing , that case related to having one version of nuget package and another version of the runtime.

**Comment by racingcow** [comment]
I've always had similar issues in VS2022. I get "Unable to evaluate" in most all object watches, quick watches, or hovering mouse over the code, such as this generic list...

![image](https://user-images.githubusercontent.com/1176637/200023853-3ff66ec8-d6ad-49f4-9e55-f14d7c190735.png)

**Comment by thaystg** [comment]
Which version of Microsoft.AspNetCore.Components.WebAssembly.DevServer nuget package are you using?

**Comment by ehsangfl** [comment]
@thaystg  all version from dotnet 5 since 6.0.10
this is not happen on simple "counter app"
this happens on a solution with multiple dll and rcl

**Comment by thaystg** [comment]
Can you share with us a core where I can reproduce it?

**Comment by racingcow** [comment]
Sorry for the delay. Microsoft.AspNetCore.Components.WebAssembly.DevServer version 6.0.3 for me. I'll see if I can get a working sample together. As @ehsangfl mentioned, it does seem to just manifest on larger projects.

**Comment by Nexusmeister** [comment]
Got the same issue while evaluating the `WebAssemblyHostConfiguration`
![image](https://github.com/mono/mono/assets/37147173/6fd57840-8df2-4e17-a78b-d665a6b3b150)

I got the following dependencies on .NET 7.0.302:
```xml
<PackageReference Include="Microsoft.AspNetCore.Components.WebAssembly" Version="7.0.5" />
<PackageReference Include="Microsoft.AspNetCore.Components.WebAssembly.DevServer" Version="7.0.5" PrivateAssets="all" />
<PackageReference Include="Microsoft.Extensions.Configuration.UserSecrets" Version="7.0.0" />

```

**Comment by arizvi-techgies** [comment]
This is the same for me. I'm using ".NET 7" Maui Blazor not able to debug

**Comment by lewing** [comment]
please open an issue in dotnet/runtime for any issue in .NET 5 or above

**Comment by lewing** [comment]
.NET wasm runtimes have followed the public .NET support schedule since .NET 5.0 and there are no supported releases of browser-wasm at this point.

**Comment by ilonatommy** [comment]
> Got the same issue while evaluating the `WebAssemblyHostConfiguration`

![image](https://github.com/mono/mono/assets/32700855/d2976c39-93ff-4580-92a1-7898ed2f5f1c)

Cannot reproduce with this version. On creating a new issue, as Larry mentioned, please add a reproduction project if possible.
